ENH: Raise ValueError for unsupported Window functions

ihsansecer opened the pull request:
- [x] ref #26597
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Weighted std could be implemented but there are also other functions weighted `Window` is lacking.

ihsansecer committed:
ENH: Raise ValueError for unsupported Window functions

ihsansecer committed:
Improve docs for win_type

ihsansecer committed:
Add whatsnew

jreback reviewed (changes requested):
this seems like overkill 

when you create a Window it already has restricted methods

what error message does this have now (before this PRA)?

ihsansecer commented:
@jreback it is ok when directly calling an unsupported function but raised error (`AttributeError: 'Window' object has no attribute 'dtype'`) is not informative for `agg` function

jreback commented on pandas/core/window.py:
```diff
@@ -659,6 +660,26 @@ def validate(self):
             raise ValueError("Invalid window {0}".format(window))
 
+    def _validate_aggregate(self, arg, *args, **kwargs):
+        # supported weighted window functions
```
rather that this (listing methods)

you can look a bit more into where the methods are called and catch an invokation error (AttributError) for example

we already have quite a bit of code to facilitate this and don’t want to add more layers

ihsansecer committed:
Catch AttributeError and raise ValueError instead

ihsansecer commented on pandas/core/base.py:
```diff
@@ -603,9 +609,6 @@ def _aggregate_multiple_funcs(self, arg, _level, _axis):
                     keys.append(col)
                 except (TypeError, DataError):
                     pass
-                except ValueError:
```
I am not sure what this is for just checked some tests locally and didn't fail

WillAyd replied:
Hmm a little wary here - this does show as hit on code coverage:

https://codecov.io/gh/pandas-dev/pandas/src/master/pandas/core/base.py#L608

ihsansecer replied:
It seems this is for nuisance columns

ihsansecer committed:
Hangle list-like args

WillAyd commented on pandas/core/base.py:
```diff
@@ -314,7 +314,13 @@ def _try_aggregate_string_function(self, arg, *args, **kwargs):
+                    "'{cls}' object".format(arg=arg, cls=type(self).__name__)
+                )
 
         raise ValueError("{arg} is an unknown string function".format(arg=arg))
```
Does this get hit any more?

ihsansecer replied:
It does for single string argument or a dictionary. Other one is for list or tuple with invalid functions

ihsansecer committed:
Skip if no scipy

ihsansecer committed:
Format using black

jreback commented on pandas/core/base.py:
```diff
@@ -554,7 +560,19 @@ def is_any_frame():
+                    None,
+                )
+
+            # if no results
```
can you give a more comprehensive comment here.

Why are we raising ValueError (and not AttributeError from the inner function)?

ihsansecer replied:
I guess AttributeError would be a better option if not inconvenient since ValueErrors are conflicting. Will change it

jreback reviewed (changes requested):
looks good, question about if we are correctly raising a ValueError in the list-like case (which you are catching).

ihsansecer committed:
Raise AttributeError instead of conflicting ValueError

ihsansecer committed:
Merge remote-tracking branch 'upstream/master' into val-window-agg

ihsansecer committed:
Remove ' from whatsnew

ihsansecer commented:
Please don't merge this as it raises `ValueError: no results` in some cases. Will check it when I have time

ihsansecer committed:
Change ValueError with AttributeError and catch TypeError

AttributeError is raised since it is conflicting with the one when list-like arg is provided. TypeError is catched since it is raised when 'median' is passed as an arg

ihsansecer committed:
Test each function one by one

ihsansecer committed:
Test unknown function

ihsansecer committed:
Format

WillAyd reviewed (approved):
lgtm ex any @jreback comments and any CI failures

ihsansecer committed:
Merge remote-tracking branch 'upstream/master' into val-window-agg

ihsansecer commented:
`skew` and `kurt` functions as an argument to `agg` raises `unknown string function` error. `min` and `max` functions on the other hand doesn't raise anything and returns `Window`object.

jreback commented on pandas/core/base.py:
```diff
@@ -314,9 +314,16 @@ def _try_aggregate_string_function(self, arg, *args, **kwargs):
+                return f(self, *args, **kwargs)
+
+            except (AttributeError, TypeError):
+                raise AttributeError(
```
I would pass here, then just use a single raise (line 326)

jreback commented on pandas/core/window.py:
```diff
@@ -511,6 +511,7 @@ class Window(_Window):
         Set the labels at the center of the window.
     win_type : str, default None
         Provide a window type. If ``None``, all points are evenly weighted.
+        Other types are only applicable for `mean` and `sum` functions.
```
this *should* be on .agg so I wouldn't add anything here

ihsansecer replied:
`agg` is using `shared_docs`. I was avoiding to copy it and make a single line change. Will remove it only if no objection

ihsansecer committed:
Raise single error for invalid functions

ihsansecer committed:
Remove from doc

ihsansecer committed:
Remove new line

jreback commented:
lgtm @ihsansecer ping on green.

jreback commented:
thanks @ihsansecer

jreback commented:
btw now that pandas/tests/window/ exists would love to split test_window even more :->

ihsansecer commented:
@jreback noted :)
